src/main/java/com/hubspot/jinjava/util/ChunkResolver.java

```diff
@@ -0,0 +1,355 @@
+        prevChar = c;
+        tokenBuilder.append(c);
+        tokenBuilder.append(resolveChunk(String.join("", getChunk(c))));
+        tokenBuilder.append(prevChar);
+        continue;
+      } else if (isTokenSplitter(c)) {
+        prevChar = c;
+
+        miniChunkBuilder.append(resolveToken(tokenBuilder.toString()));
+        tokenBuilder = new StringBuilder();
+        if (isMiniChunkSplitter(c)) {
+          chunks.add(resolveChunk(miniChunkBuilder.toString()));
+          chunks.add(String.valueOf(c));
+          miniChunkBuilder = new StringBuilder();
+        } else {
+          miniChunkBuilder.append(c);
+        }
+        continue;
+      }
+      prevChar = c;
+      tokenBuilder.append(c);
+    }
+    miniChunkBuilder.append(resolveToken(tokenBuilder.toString()));
+    chunks.add(resolveChunk(miniChunkBuilder.toString()));
+    return chunks;
+  }
+
+  private boolean isTokenSplitter(char c) {
+    return (!Character.isLetterOrDigit(c) && c != '_' && c != '.');
+  }
+
+  private boolean isMiniChunkSplitter(char c) {
+    return c == ',';
+  }
+
+  private String resolveToken(String token) {
+    if (StringUtils.isBlank(token)) {
+      return "";
+    }
+    try {
+      String resolvedToken;
+      if (WhitespaceUtils.isQuoted(token) || RESERVED_KEYWORDS.contains(token)) {
+        resolvedToken = token;
+      } else {
+        Object val = interpreter.retraceVariable(
+          token,
+          this.token.getLineNumber(),
+          this.token.getStartPosition()
+        );
+        if (val == null) {
+          try {
+            val = interpreter.resolveELExpression(token, this.token.getLineNumber());
+          } catch (UnknownTokenException e) {
+            // val is still null
+          }
+        }
+        if (val == null) {
+          resolvedToken = token;
+        } else {
+          resolvedToken = getValueAsJinjavaString(val);
+        }
+      }
+      return resolvedToken.trim();
+    } catch (DeferredValueException | JsonProcessingException e) {
+      deferredWords.addAll(findDeferredWords(token));
+      return token.trim();
+    }
+  }
+
+  // Try resolving the chunk/mini chunk as an ELExpression
+  private String resolveChunk(String chunk) {
+    if (StringUtils.isBlank(chunk)) {
+      return "";
+    } else if (RESERVED_KEYWORDS.contains(chunk)) {
+      return chunk;
+    }
+    try {
+      String resolvedChunk;
+      Object val = interpreter.resolveELExpression(chunk, token.getLineNumber());
+      if (val == null) {
+        resolvedChunk = chunk;
+      } else {
+        resolvedChunk = getValueAsJinjavaString(val);
+      }
+      return resolvedChunk.trim();
+    } catch (Exception e) {
+      deferredWords.addAll(findDeferredWords(chunk));
+      return chunk.trim();
+    }
+  }
+
+  public static String getValueAsJinjavaString(Object val)
+    throws JsonProcessingException {
+    return OBJECT_MAPPER
+      .writeValueAsString(val)
+      .replaceAll("(?<!\\\\)(?:\\\\\\\\)*(\\\\n)", "\n")
```

Comment:
what do these do?

Reply:
These swap out a couple of characters from the JSON string. Double quotes are replaced with single quotes, and an escaped `n` is replaced with the newline character, and an escaped `"` is replaced with a double-quote character.

The `n` and `"` replacements are because if they don't get replaced, when the JSON string is output, it would have literal `\n`s and `\"`s in it where we actually would want newlines and `"`s

Reply:
OK, adding a comment here would be helpful for non-obvious code like this.